Feature/add-urth-pricefeed

krisoshea-eth opened the pull request:
This PR adds the URTH synth Polygon API price feed to this repo to allow the Liquidator and Disputer bots to fetch URTH price data. This appears to be working but requires further testing / review from team to confirm if the bots are fetching the correct data. One issue that I need to fix before merging is to update the URL such that the bot can fetch a custom date. I am still looking into how to do this but I wanted to confirm that the bot is fetching the data correctly first before making another change to the code.

krisoshea-eth committed:
Update to include URTHPriceFeed.ts

Updated this branch to add URTH price feed to the CreatePriceFeed.ts file

krisoshea-eth committed:
Update URTHPriceFeed.ts

Removed unused code from this file in order to remove errors

krisoshea-eth committed:
Update CreatePriceFeed.ts

Removes unnecessary param from URTH-api else-if block

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/URTHPriceFeed.ts:
```diff
@@ -0,0 +1,192 @@
+import { BN } from "../types";
```
I would suggest changing the name from `URTHPriceFeed.ts` to `PolygonApiPriceFeed.ts`. Also, changing URTH to PolygonApi elsewhere as well e.g. in config.type. That is more apt, and shows our intent of using this price feed for other synths in future.

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/URTHPriceFeed.ts:
```diff
@@ -0,0 +1,192 @@
+
+    // 1. Construct URL.
+    // See https://polygon.io/docs/stocks/getting-started.
+    const url = "https://api.polygon.io/v1/open-close/" + this.index + `2023-01-09?adjusted=true&apiKey=${this.apiKey}`;
```
The way the update function works is a bit complex. It basically runs at regular intervals, decided by the `minTimeBetweenUpdates` variable. The update function then get's a historical response from the API which is between 2 timestamps decided by `startDateString` and `endDateString`. The start date depends on the `lookback` variable.

The current price is stored in the `currentPrice` variable which is retrieved from the historical prices that the API sent back.

In the above code, the date is hardcoded to `2023-01-09` which is incorrect, it should instead use the dates I explained above.

KohliSuraj replied:
You can take inspiration on how to implement above from `CommoditiesApiPriceFeed.ts` file.

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/URTHPriceFeed.ts:
```diff
@@ -0,0 +1,192 @@
+    this.lastUpdateTime = currentTime;
+  }
+
+  public getCurrentPrice(): BN | null {
```
You can test that this function is returning the correct Price when the bot is run..

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/URTHPriceFeed.ts:
```diff
@@ -0,0 +1,192 @@
+    return this.currentPrice;
+  }
+
+  public async getHistoricalPrice(time: number, ancillaryData?: string, verbose?: boolean): Promise<BN | null> {
```
I'm not too sure when this function is called and how to test this. It's used by the bots, to figure out historical prices.

KohliSuraj committed:
getting historical prices for URTH

krisoshea-eth committed:
Change URTHPricesFeed.ts name to PolygonApiPriceFeed.ts

Also, changes URTH to PolygonApi elsewhere, where appropriate

KohliSuraj committed:
URTH to use TwelveData API

krisoshea-eth committed:
Update TwelveDataApiPriceFeed.ts

Update TwelveDataApiPriceFeed.ts naming conventions

krisoshea-eth committed:
Update TwelveDataApiPriceFeed.ts

-Updated this file to oldestHistoricalPrice function which fetches the price data from the last data set in the array returned by the api response object.
Fixed the currentPrice function to accurately fetch the current price data by returning the price data for the first data set in the array returned by the Api response object.
-Altered the getHistoricalPrice function to fetch the correct price data based on the array that orders price data from newest to oldest (this needs testing)
-

KohliSuraj committed:
Revert "Update TwelveDataApiPriceFeed.ts"

This reverts commit 619810cf93a576c3f5f9d67816863acc618564d4.

KohliSuraj committed:
update TwelveData API for URTH to get closingPrice

KohliSuraj committed:
logging timezone
